src/upgrades/mongodb_upgrade.py

```diff
@@ -0,0 +1,147 @@
+            self._upgrade.upgrade_resumed = False
+            # Only call `_reconcile_upgrade` on leader unit to avoid race conditions with
+            # `upgrade_resumed`
+            self._reconcile_upgrade()
+
+    def _on_resume_upgrade_action(self, event: ActionEvent) -> None:
+        if not self.charm.unit.is_leader():
+            message = f"Must run action on leader unit. (e.g. `juju run {self.app.name}/leader {upgrade.RESUME_ACTION_NAME}`)"
+            logger.debug(f"Resume upgrade event failed: {message}")
+            event.fail(message)
+            return
+        if not self._upgrade or not self._upgrade.in_progress:
+            message = "No upgrade in progress"
+            logger.debug(f"Resume upgrade event failed: {message}")
+            event.fail(message)
+            return
+        self._upgrade.reconcile_partition(action_event=event)
+
+    def _on_force_upgrade_action(self, event: ActionEvent) -> None:
+        if not self._upgrade or not self._upgrade.in_progress:
+            message = "No upgrade in progress"
+            logger.debug(f"Force upgrade event failed: {message}")
+            event.fail(message)
+            return
+        if not self._upgrade.upgrade_resumed:
+            message = f"Run `juju run {self.app.name}/leader resume-upgrade` before trying to force upgrade"
+            logger.debug(f"Force upgrade event failed: {message}")
+            event.fail(message)
+            return
+        if self._upgrade.unit_state != "outdated":
+            message = "Unit already upgraded"
+            logger.debug(f"Force upgrade event failed: {message}")
+            event.fail(message)
+            return
+        logger.debug("Forcing upgrade")
+        event.log(f"Forcefully upgrading {self.unit.name}")
+        self._upgrade_opensearch_event.emit(ignore_lock=event.params["ignore-lock"])
+        event.set_results({"result": f"Forcefully upgraded {self.unit.name}"})
+        logger.debug("Forced upgrade")
+
+    # END: Event handlers
+
+    # BEGIN: Helpers
+    def _set_upgrade_status(self):
+        # In the future if we decide to support app statuses, we will need to handle this
+        # differently. Specifically ensuring that upgrade status for apps status has the lowest
+        # priority
+        if self.charm.unit.is_leader():
+            self.charm.app.status = self._upgrade.app_status or ActiveStatus()
+
+        # Set/clear upgrade unit status if no other unit status - upgrade status for units should
+        # have the lowest priority.
+        if isinstance(self.charm.unit.status, ActiveStatus):
+            self.charm.unit.status = self._upgrade.get_unit_juju_status() or ActiveStatus()
+
+    # END: helpers
+
+    # BEGIN: properties
+    @property
+    def _upgrade(self) -> Optional[machine_upgrade.Upgrade]:
+        try:
+            return machine_upgrade.Upgrade(self.charm)
+        except upgrade.PeerRelationNotReadyError:
+            pass
```

Comment:
```suggestion
        except upgrade.PeerRelationNotReady:
            pass
```
nit: please avoid changing API of upgrade.py/machine_upgrade.py (will make it more difficult to switch to shared code in future)

Reply:
@carlcsaposs-canonical This fails my linter (I'm suprised it doesn't fail OpenSearch's or MySQL's, custom exceptions should always end with `Error`)

Reply:
https://github.com/canonical/mysql-router-k8s-operator/blob/50b19ac9060f66d9447bbc77a10ed8ff41df3c0f/pyproject.toml#L104

imo, exceptions should not always end with `Error` because they're not always errors

Reply:
I can support that line of thinking. Thanks for the suggestion Carl, I will revert this and add that to my pyproject.toml <3